File: Genetic/src/main/scala/baldwin/BaldwinBit.scala
package baldwin

import scala.util.Random

abstract class BaldwinBit {
  def isQuestionMark: Boolean

  def toString: String
}

case class DataBit(bit: Bit) extends BaldwinBit {
  override def isQuestionMark: Boolean = false

  override def toString: String = bit.toString
}

case object QuestionMark extends BaldwinBit {
  override def isQuestionMark: Boolean = true

  override def toString: String = "?"
}

abstract class Bit {
  def toString: String
}

case object One extends Bit {
  override def toString: String = "1"
}

case object Zero extends Bit {
  override def toString: String = "0"
}

object Bit {
  def fromBool(b: Boolean): Bit = {
    if (b)
      One
    else
      Zero
  }

  def genRandom(rnd: Random): Bit = {
    fromBool(rnd.nextBoolean())
  }
}

File: Genetic/src/main/java/genetic/baldwin/BaldwinBit.java
package genetic.baldwin;

import java.util.Random;

public class BaldwinBit {
    public static final byte Zero = 0;
    public static final byte One = 1;
    public static final byte QuestionMark = 2;

    public static byte genBit(Random rand) {
        // XorShift's native method is nextLong
        return (byte) (rand.nextLong() & 1);
    }

    public static byte genBaldwinBit(Random rand) {
        if (rand.nextBoolean()) {
            return QuestionMark; // Question Mark
        } else {
            return genBit(rand);
        }
    }

}
